File: content/atlas/source/includes/fts/search-with-unionwith/basic-query.kt
import com.mongodb.kotlin.client.coroutine.MongoClient
import kotlinx.coroutines.runBlocking
import org.bson.Document

fun main() {
    // connect to Atlas cluster
    val uri = "<connection-string>"
    val mongoClient = MongoClient.create(uri)

    // set namespace
    val database = mongoClient.getDatabase("sample_training")
    val collection = database.getCollection<Document>("companies")

    runBlocking {
        // define pipeline
        val pipeline1 = listOf(
            Document("\$search", Document("text",
                Document("query", "Mobile")
                    .append("path", "name"))), Document("\$project", Document("score",
                Document("\$meta", "searchScore"))
                .append("_id", 0)
                .append("number_of_employees", 1)
                .append("founded_year", 1)
                .append("name", 1)), Document("\$set", Document("source", "companies")),
            Document("\$limit", 3)
        )

        val pipeline2 = listOf(
            Document(
                "\$search", Document(
                    "text",
                    Document("query", "Mobile")
                        .append("path", "business_name")
                )
            ),
            Document("\$set", Document("source", "inspections")),
            Document(
                "\$project", Document(
                    "score",
                    Document("\$meta", "searchScore")
                )
                    .append("source", 1)
                    .append("_id", 0)
                    .append("business_name", 1)
                    .append("address", 1)
            ),
            Document("\$limit", 3),
            Document("\$sort", Document("score", -1))
        )

        val unionWithStage: MutableList<Document> = ArrayList()
        val unionWith = Document(
            "\$unionWith", Document("coll", "inspections")
                .append("pipeline", pipeline2)
        )
        unionWithStage.add(unionWith)
        val finalPipeline: MutableList<Document> = ArrayList(pipeline1)
        finalPipeline.addAll(unionWithStage)

        // run pipeline and print results
        val resultsFlow = collection.aggregate<Document>(finalPipeline)
        resultsFlow.collect { println(it) }

    }
    mongoClient.close()
}

File: content/atlas/source/includes/fts/search-with-unionwith/basic-query.cs
using MongoDB.Bson;
using MongoDB.Driver;
using MongoDB.Driver.Search;

public class Program
{
    public static void Main(string[] args)
    {
        // connect to your Atlas cluster
        string connectionString = "<connection-string>";
        var client = new MongoClient(connectionString);
        
        // define namespace
        var database = client.GetDatabase("sample_training");
        var collection = database.GetCollection<BsonDocument>("companies");
        
        // define pipeline stage
        var searchStage1 = new BsonDocument("$search", new BsonDocument{{ "text", new BsonDocument
            {{ "query", "Mobile" },{ "path", "name" }}
        }});
        var projectStage1 = new BsonDocument("$project", new BsonDocument{
            { "score", new BsonDocument("$meta", "searchScore") },
            { "_id", 0 },{ "number_of_employees", 1 },{ "founded_year", 1 },{ "name", 1 }
        });
        var setStage1 = new BsonDocument("$set", new BsonDocument{{ "source", "companies" }});
        var limitStage1 = new BsonDocument("$limit", 3);
        
        // define subpipeline
        var searchStage2 = new BsonDocument("$search", new BsonDocument{{ "text", new BsonDocument
            {{ "query", "Mobile" },{ "path", "business_name" }}
        }});
        var setStage2 = new BsonDocument("$set", new BsonDocument{ { "source", "inspections" } });
        var projectStage2 = new BsonDocument("$project", new BsonDocument{
            { "score", new BsonDocument("$meta", "searchScore") }, 
            { "source", 1 }, { "_id", 0 }, { "business_name", 1 }, { "address", 1 }
        });
        var limitStage2 = new BsonDocument("$limit", 3);
        var sortStage2 = new BsonDocument("$sort", new BsonDocument{{ "score", -1 }});
        var unionWithPipeline = new List<BsonDocument>{searchStage2, setStage2, projectStage2, limitStage2, sortStage2};
        var unionWithStage = new BsonDocument("$unionWith", new BsonDocument
        {
            { "coll", "inspections" },
            { "pipeline", new BsonArray(unionWithPipeline) }
        });
        var aggregationPipeline = new List<BsonDocument> {searchStage1, projectStage1, setStage1, limitStage1,unionWithStage};
        
        // run pipeline
        var result = collection.Aggregate<BsonDocument>(aggregationPipeline).ToList();
        
        //print results
        foreach (var document in result)
        {
            Console.WriteLine(document);
        }
    }
}
